Library/Homebrew/formula_installer.rb

```diff
@@ -1182,9 +1182,11 @@ def fetch
     if pour_bottle?(output_warning: true)
       formula.fetch_bottle_tab
     else
-      if formula.core_formula? && Homebrew::EnvConfig.install_from_api?
-        url = "https://raw.githubusercontent.com/#{formula.tap.full_name}/#{formula.tap_git_head}/Formula/#{formula.name}.rb"
-        @formula = Formulary.factory(url, formula.active_spec_sym,
+      if formula.class.loaded_from_api
+        resource = formula.resource("ruby-source")
```

Comment:
I think this would be a good time to unify the formula and cask download logic. Making this a new resource feels pretty strange.

Reply:
For formulae, it actually makes a lot of sense. Everything in a formula _is_ a resource. Original source archives are a resource, bottles are a resource, patches are a resource. The strange bit here might be the DSL I guess.

It fits in very very well. In formula installer the same output progress information on downloads etc will be used for this. The whole download system in formulae will all be using the same system - I didn't need to do anything here to check checksums etc as its all already there.

It's the cask side I think that needs a big refactor. It has zero caching, downloads source files more so than strictly necessary, has zero progress information or feedback, etc. I was going to do so for this PR, but it seemed to large to fit bundle into the same PR. Will take a look now at what can be done.

Reply:
> Everything in a formula is a resource.

Yeh but arguably: this is the formula itself rather than a resource on the formula.

> Using the resource system fits in very very well.

If this was a resource that:
- wasn't nested on formula
- could be used by casks

that would make more sense to me

> Will take a look now at what can be done.

Thanks 🙇🏻

I just think it'll be easier to ensure that this works and behaves consistently if we're consolidating that logic first.

Reply:
> If this was a resource that:
> 
> * wasn't nested on formula
> * could be used by casks

Yeah I'm hoping we could use the resource system in casks too. Will experiment with that a bit (separately).

Reply:
Yeah, might be a case of a `Downloadable` base class (see also: `Cask::Download`).